node/src/chain/acceptor.rs

```diff
@@ -604,7 +580,133 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
+        };
+        let mut finalized_blocks = BTreeMap::new();
+
+        let current_height = blk.header().height;
+        let mut labels = BTreeMap::new();
+
+        // Retrieve latest blocks up to the last finalized
+        let mut old_final_hash = None;
+        for height in (0..current_height).rev() {
+            let (hash, label) = db.fetch_block_label_by_height(height)?.ok_or(
+                anyhow!("Cannot find block label for height {height}"),
+            )?;
+            if let Label::Final(_) = label {
+                old_final_hash = Some(hash);
+                break;
+            }
+            labels.insert(height, (hash, label));
+        }
+        let old_final_hash =
+            old_final_hash.expect("Unable to find last finalized block hash");
+        let old_final_state_root = db
+            .fetch_block_header(&old_final_hash)?
+            .ok_or(anyhow!(
+                "Cannot get header for last finalized block hash {}",
+                to_str(&old_final_hash)
+            ))?
+            .0
+            .state_hash;
+
+        // Iterate from tip to last_final to set label::confirmed
+        let mut stable_count = 1;
+        for (&height, (hash, label)) in labels.iter_mut().rev() {
+            match label {
+                Label::Accepted(ref confirmed_after) => {
+                    if confirmed_after >= &stable_count {
+                        *label = Label::Confirmed(current_height - height);
+                        db.store_block_label(height, hash, *label)?;
+                        stable_count += 1;
+                    } else {
+                        break;
+                    }
+                }
+                Label::Attested(ref confirmed_after) => {
+                    if confirmed_after >= &stable_count {
+                        *label = Label::Confirmed(current_height - height);
+                        db.store_block_label(height, hash, *label)?;
+                        stable_count += 1;
+                    } else {
+                        break;
+                    }
+                }
+                Label::Confirmed(_) => {
+                    stable_count += 1;
+                    continue;
+                }
+                Label::Final(_) => {
+                    warn!("Found a final block during rolling finality scan. This should be a bug");
+                    break;
+                }
+            }
+        }
+
+        // Iterate from last_final to tip to set label::final
+        for (height, (hash, mut label)) in labels.into_iter() {
+            match label {
+                Label::Final(_) => {
+                    warn!("Found a final block during rolling finality. This should be a bug")
+                }
+                Label::Accepted(_) => break,
+                Label::Attested(_) => {
+                    warn!("Found an attested block during rolling finality. This should be a bug")
```

Comment:
I'm not sure finding an `Attested` is a bug. 
If followed by an `Accepted` that has not been confirmed, it would still be in `labels`.

E.g.: we have
5:FINAL - 6:ATTESTED - 7:ACCEPTED(3) - 8:ATTESTED - 9:`blk`
`blk` is ATTESTED and confirmes 8 but not 7. The previous loop breaks at 7, but 6 is still in `labels`. So here we would start the loop with 6, which is ATTESTED (but not confirmed).

I think we should break here

Reply:
You're simply right